### PR TITLE
all of it: add a persist sink/source

### LINKED ISSUE #11905: Add a source type that reads out of persist from compute
We currently render sources every time they are requested and storage sends all data over a TCP connection to whoever needs it. Once storage writes all source data into a `persist` collection we can replace the read-side with just reading from `persist` directly.

### PR DESCRIPTION
**Pull request by aljoscha**
### Motivation

This unblocks future work and adds at least one desirable feature. This unblocks storage/compute separation because it will eventually allow sources to write into persist, and compute will then read out of persist using a persist source.

As of now, this enables the following user interactions:

```
> CREATE TABLE foo (name TEXT);

> CREATE SINK persist_sink FROM foo INTO PERSIST
  CONSENSUS 'sqlite:///tmp/mz-persist/consensus'
  BLOB 'file:///tmp/mz-persist/blob'
  SHARD 'sa3150c18-f7a0-4062-b189-463531b9f36b'
  ENVELOPE NONE;

> CREATE SOURCE persist_source FROM PERSIST
  CONSENSUS 'sqlite:///tmp/mz-persist/consensus'
  BLOB 'file:///tmp/mz-persist/blob'
  SHARD 'sa3150c18-f7a0-4062-b189-463531b9f36b'
  (text string);

materialize=> INSERT INTO foo VALUES ('ciao bello 5');
INSERT 0 1

materialize=> SELECT * FROM persist_source AS OF NOW();
     name     
--------------
 ciao bello 5
(1 row)

materialize=> CREATE MATERIALIZED VIEW view AS SELECT * FROM persist_source;
CREATE VIEW
materialize=> SELECT * FROM view;
     name     
--------------
 ciao bello 5
(1 row)
```

This is somewhat "loopy", because a persist sink also presents as a source, that can be queried, but it seemed quite natural to me. There are some rough edges in the implementation, more below.

Also, this treats the persist `Client` as a poor-man's storage client, and both the sink and source can only handle a single persist shard. In the future, a storage abstraction will hide potentially multiple shards behind a single storage collection, and the sink/source will need to be changed to accommodate that.

### Motivation

This PR adds a known-desirable feature: #11905.

Fixes #11905.

### Tips for reviewer

There are multiple parts in one big commit and some preliminary cleanup commits.

Outline of the changes:
 - Adds source/sink implementations for writing to and reading from persist.
 - Adds DDL for creating a `PERSIST SINK`.
 - Massages the compute and storage controllers to treat a persist sink both as a sink and a source. The sink creates a collection that we can read from in peeks and materialized views.
 - There is no explicit `PERSIST SOURCE`, these are created by reading from a persist sink. In the future, the same underlying parts will be used for reading from a collection that is created by ingesting an external source.

#### Caveats / Future Work

There are many `TODOs` in the code itself. There are some bigger things that we might nor might not want to tackle immediately:

 - ~~Is "reading from a sink" a good thing?~~ NO
 - ~~Who should be responsible for maintaining the collection state of the sink? The compute or storage controller? On `main`, compute is responsible for sinks. My code changes that slightly, because we want to allow reading from the sink, which is easier if it is maintained by the storage controller.~~ resolved
 - ~~We should probably make collections a first-class thing and separate them from the sink that spawns them. But that requires more work. What should be the lifecycle of the collection spawned by the sink?~~
 - We don't know what to do with errors that occur in the sink. Currently we panic. We have similar problems with all our sinks.
 - There are hardcoded timeouts for all the persist API calls. Should we thread through a timeout configuration? Should persist calls maybe not have a timeout and keep trying infinitely?
 - The sink has `blob` and `consensus` parameters. These should not be there but instead be picked up from STORAGE/`storaged` configuration.
 - The sink and source take `SHARD` parameters, which are `s<UUID>`. We could (should!) change this to be some human-readable collection identifier. Also, someone needs to think about the lifecycle/management of collections.
 - If we maintain information about collections somewhere, that should also make it possible to determine the schema of a persist source, instead of requiring the user to manually specify columns, as we do now.

My preference would be to merge this early and have someone iterate on it, but as it is things are somewhat alpha quality so idk... 🤷‍♂️ 

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### CODE REVIEW

**Comment by frankmcsherry** [comment]
I have a bunch of questions about the design, in particular of merging the sink and source. At the moment, sinks are a cluster-local thing that impose compaction constraints on their sources. If they are "the same thing" it seems like this might introduce cross-cluster constraints, where a reader of a source needs to prevent compaction back through the source, to the sink, in a different cluster, which might read from a source that comes from a sink on another clusters, etc.

Some amount of this is easier for me to grok if sinks write at a persist location, and sources read from a persist location, and they are not otherwise observed to be related. Is there a reason to avoid that design / what are the trade-offs?

**Comment by aljoscha** [comment]
> Is there a reason to avoid that design / what are the trade-offs?

It required less code, and it avoids an extra `CREATE SOURCE ...` statement from the user. But I also have a version of this that has an explicit `CREATE SOURCE ...` that I was planning to dust off. Now that I see the hacks I had to apply to the compute/storage controllers I think that might be the better solution as well.

File: src/compute/src/render/persist_source.rs
Comment (guswynn) [inline src/compute/src/render/persist_source.rs:53]: curious, does rustfmt work inside this macro?

Comment (aljoscha) [reply src/compute/src/render/persist_source.rs:53]: it does not!

**Comment by aljoscha** [comment]
I updated this to make persist sources explicit. The description is updated to show what user interactions are now possible.

File: src/compute/src/render/mod.rs
Comment (frankmcsherry) [inline src/compute/src/render/mod.rs:129]: Nit, but the intended separation of responsibilities would probably put this in STORAGE, like the other `ComputeReplay` implementors. Like, if STORAGE wants to change how the implementation of this source works, they should be able to and COMPUTE should just have to link against their code.

File: src/compute/src/render/persist_source.rs
Comment (frankmcsherry) [inline src/compute/src/render/persist_source.rs:61]: Should the method perhaps instead take a `Vec<ShardId>`, which we can either rig to be non-empty for worker zero only, or to reflect whatever we eventually decide?

File: src/compute/src/render/persist_source.rs
Comment (frankmcsherry) [inline src/compute/src/render/persist_source.rs:109]: I can't really tell what any of this does. That may be fine if everyone else gets it, but I bet some comments would help generally!

File: src/compute/src/render/persist_source.rs
Comment (frankmcsherry) [inline src/compute/src/render/persist_source.rs:124]: The timely source should not have a `Capability` and should instead have only `CapabilitySet` types. We can fix that.

File: src/compute/src/render/persist_source.rs
Comment (frankmcsherry) [inline src/compute/src/render/persist_source.rs:161]: What does `None` mean in this context? Downgrading the capability to `[]` seems to be something pretty strong, not just e.g. "connection severed" or the like. It seems like maybe `None` when the frontier is not `[]` could/should mean "disconnection error"?

File: src/compute/src/sink/mod.rs
Comment (frankmcsherry) [inline src/compute/src/sink/mod.rs:13]: Eventually I think we'll want to put these over in STORAGE, but for the time being we cannot (TAIL being a SINK complicates that). Nothing to do at the moment here!

File: src/compute/src/sink/persist_sink.rs
Comment (frankmcsherry) [inline src/compute/src/sink/persist_sink.rs:65]: I'd maybe toss an explicit `// TODO` here, as I think we'll want to sort this out somewhat promptly. Like, no reason afaict that we shouldn't just be writing to one persist shard per worker.

File: src/coord/src/coord/dataflow_builder.rs
Comment (frankmcsherry) [inline src/coord/src/coord/dataflow_builder.rs:170]: What's going on here? Toss a comment in, so that future maintainers can grok what needs doing.

File: src/dataflow-types/src/client/controller.rs
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller.rs:37]: This was added as a requirement to lots of generic arguments, but I couldn't see where it was required. Can you explain?

File: src/dataflow-types/src/client/controller/storage.rs
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller/storage.rs:280]: I have a bunch of thoughts here that might need to be follow-on work. Mostly of the form "we want to find a way to have the controller contain less of this sort of logic itself, and instead shell out to implementations elsewhere". Otherwise the controllers will become the new `coord.rs`.

File: src/dataflow-types/src/client/controller/storage.rs
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller/storage.rs:721]: We should discuss these! It seems like maybe there is a missing abstraction for what there are needed for. Ideally we could bundle this up in an abstraction that presents as what it is needed for, vs "I might need these fields in some cases".

**Review by frankmcsherry — APPROVED** [review]
I read everything and left several comments. From the COMPUTE side of things it checks out. I had some thoughts about how STORAGE might want to organize itself to better manage its complexity, which may be unwelcome I'm not sure. :)

I also have some questions about the persist source and which envelope it has, as this will potentially break e.g. monotonicity reasoning if it is `None`. I wasn't able to see where this gets set, but my eyes glazed over a bit reading the SQL side of things.

File: src/compute/src/render/mod.rs
Comment (aljoscha) [reply src/compute/src/render/mod.rs:129]: That's a good point! I'll move the code to the `storage` crate. It would also be nice to "hide" it inside the boundary but that would require implementing code in all the `ComputeReplay` impls (of which there is only one real one) because the boundary/replay impls are orthogonal to replay via persist. Plus, in the long run the current `ComputeReplay` impls will go away and we will have only persist replay.

I can't just add a `PersistReplay` right now because the boundary is not instantiated per source but is global to the compute instance. Is that what you had in mind?

File: src/compute/src/render/persist_source.rs
Comment (aljoscha) [reply src/compute/src/render/persist_source.rs:61]: That's an excellent point! It should be like this, but making the source behave nicely when reading from multiple shards will require some elbow grease. Is it alright if I outline the required changes as a TODO and merge as is?

File: src/compute/src/render/persist_source.rs
Comment (aljoscha) [reply src/compute/src/render/persist_source.rs:109]: I'll add more comments explaining this.

File: src/compute/src/render/persist_source.rs
Comment (aljoscha) [reply src/compute/src/render/persist_source.rs:124]: Should we do that before merging? I seem to recall you had a draft PR that did that, right?

File: src/compute/src/render/persist_source.rs
Comment (aljoscha) [reply src/compute/src/render/persist_source.rs:161]: It means that the stream (which we're setting up above) is exhausted, but that should never happen because we keep polling for new updates from persist forever in the asyn block above. I'll change this to panic if the stream is ever exhausted.

File: src/dataflow-types/src/client/controller.rs
Comment (aljoscha) [reply src/dataflow-types/src/client/controller.rs:37]: Resolved by hiding the `ReadHandle`, which was causing this behind a trait object.

Comment (aljoscha) [reply src/dataflow-types/src/client/controller.rs:37]: Resolved by hiding the `ReadHandle`, which was causing this behind a trait object.

File: src/compute/src/sink/persist_sink.rs
Comment (aljoscha) [reply src/compute/src/sink/persist_sink.rs:65]: Will put in a `TODO`

File: src/coord/src/coord/dataflow_builder.rs
Comment (aljoscha) [reply src/coord/src/coord/dataflow_builder.rs:170]: I'm rolling this change back because it's not needed right now. I had it because I introduced a `SourceConnector::Persist` variant but we should do that later, if at all.

File: src/dataflow-types/src/client/controller/storage.rs
Comment (aljoscha) [reply src/dataflow-types/src/client/controller/storage.rs]: Resolved by hiding the ReadHandle, which was causing this behind a trait object. Now there is only one `Option<CollectionReadHandle<>>` field. That will stop being optional once all sources go through persist.

((recording what I responded elsewhere, in Slack))

File: src/compute/src/render/persist_source.rs
Comment (frankmcsherry) [reply src/compute/src/render/persist_source.rs]: That sounds great; thanks!

File: src/compute/src/render/persist_source.rs
Comment (frankmcsherry) [reply src/compute/src/render/persist_source.rs]: I think feel free to merge. That work is in @cjubb39 's hands at the moment, and it isn't the most important thing it does. We can clean this up afterwards. Also @guswynn will want it too. :D

File: src/compute/src/render/persist_source.rs
Comment (frankmcsherry) [reply src/compute/src/render/persist_source.rs]: TODOs are great, and idk maybe we should file a general issue to make sure persist sources and sinks are multi-worker. But this should not prevent merging this.

File: src/dataflow-types/src/client/controller/storage.rs
Comment (aljoscha) [reply src/dataflow-types/src/client/controller/storage.rs]: I maybe addressed this by moving the code that decides to create a read handle to the source connector.

**Comment by aljoscha** [comment]
I had to rebase and force-push because some of the recent changes to persist APIs.

**Comment by aljoscha** [comment]
I had to rebase and force-push because some of the recent changes to persist APIs.

File: src/compute/src/sink/persist_sink.rs
Comment (teskje) [inline src/compute/src/sink/persist_sink.rs:180]: Out of curiosity, what is the reason for creating these `frontier1` variables?

Comment (aljoscha) [reply src/compute/src/sink/persist_sink.rs:180]: These are leftovers from when I had to `clone()` them in an earlier version of the code, I'll get rid of them

File: src/dataflow-types/src/client/controller/storage.rs
Comment (teskje) [inline src/dataflow-types/src/client/controller/storage.rs:536]: Is this a left-over?

Comment (aljoscha) [reply src/dataflow-types/src/client/controller/storage.rs:536]: Yes, this has been open for so long that I'm sure there's more. Thanks for catching these! 😃

Comment (teskje) [reply src/dataflow-types/src/client/controller/storage.rs:536]: I'm afraid I cannot contribute on a much higher level than this, but I'll continue flagging things that I notice!

File: src/dataflow-types/src/client/controller/storage.rs
Comment (petrosagg) [inline src/dataflow-types/src/client/controller/storage.rs:666]: The `Controller` type that implement `StorageController` already directly talks to `persist`. Is there a reason to hide this behind a trait?

**Review by petrosagg — APPROVED** [review]
I went through all of this and I think it's good to merge and iterate on!

File: src/dataflow-types/src/client/controller/storage.rs
Comment (aljoscha) [reply src/dataflow-types/src/client/controller/storage.rs:666]: Adding a `ReadHandle` to this structs "infects" all the types that contain it with `T: Timestamp + Lattice + Codec64` constraints. It's probably a problem in the first place that `ReadHandle` has those on the struct itself, but I thought it would be nice anyways to encapsulate.

File: src/compute/src/sink/persist_sink.rs
Comment (aljoscha) [reply src/compute/src/sink/persist_sink.rs:96]: Yes, I changed that for sources but forgot to do it for sinks. Changing it before merging.

File: src/compute/src/sink/persist_sink.rs
Comment (aljoscha) [reply src/compute/src/sink/persist_sink.rs:88]: That makes sense. I tried it quickly but I don't think I can get useful work done today. 🙈 I might have to leave this one for you.

**Pull request merged**
